stl/inc/semaphore

```diff
@@ -0,0 +1,297 @@
+// semaphore standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SEMAPHORE_
+#define _SEMAPHORE_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <semaphore> is not supported when compiling with /clr:pure.
+#endif // _M_CEE_PURE
+
+#if !_HAS_CXX20
+#pragma message("The contents of <semaphore> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+
+#include <atomic>
+#include <chrono>
+#include <limits.h>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+template <class _Rep, class _Period>
+_NODISCARD unsigned long long _Semaphore_deadline(const chrono::duration<_Rep, _Period>& _Rel_time) {
+    return __std_atomic_wait_get_deadline(
+        chrono::duration_cast<chrono::duration<unsigned long long, milli>>(_Rel_time).count());
+}
+
+template <class _Clock, class _Duration>
+_NODISCARD unsigned long _Semaphore_remaining_timeout(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+    auto _Now = _Clock::now();
+    if (_Now >= _Abs_time) {
+        return 0;
+    }
+
+    auto _Rel_time = _Abs_time - _Now;
+    static constexpr chrono::hours _Ten_days{24 * 10};
+    if (_Rel_time >= _Ten_days) {
+        return chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Ten_days).count();
+    } else {
+        return chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Rel_time).count();
+    }
+}
+
+template <ptrdiff_t _Least_max_value = INT_MAX>
+class counting_semaphore {
+public:
+    _NODISCARD static constexpr ptrdiff_t(max)() noexcept {
+        return (1ULL << (sizeof(ptrdiff_t) * CHAR_BIT - 1)) - 1;

```

Comment:
Isn't being able to set a custom size for the maximum value useless if it's not used at all? Otherwise if you accidentally release more than you acquire bugs are harder to find. I know that technically you're allowed to have it be greater than or equal to the _Least_max_value, but having it be greater than the _Least_max_value IMHO devalues its purpose, I know this is just nitpicking but this is something that is theoretically simple to change as well as helps prevent oversharing of resources as well as makes sure that no one has to innately understand the STL to find a bug.

Just a nitpick and if there's good reason to have the max() not be _Least_max_value, I'd at least like it to be documented somewhere so that some poor soul doesn't have to dig to hard to know why.

Reply:
My concern was that the check for maximum takes some machine code, more than check for negative, without being that much helpful.

Reply:
Also distinguishing cases will spawn unnecessary instantiations of the template, which may contribute to code size if barrier is not inlined. I'll go with comment.

Reply:
Ok, probably should go with that. If users are concerned about extra check penalty, they can always use defaults, with is now `numeric_limits<ptrdiff_t>max()`